### DISCUSSION TITLE
error handling

### DISCUSSION THREAD

**Discussion by kbzsl** [ropensci/targets#691] [Help]
I am planning to restructure some existing code in a targets workflow. The log and error handling is already implemented in many of existing functions (usually the NULL return value is used to signal the error).
My goal with the error handling is to execute the maximum number of targets, just skip those for which an ancestor signaled an error (this workflow generates many different results, but there are common parts). 

Reading through the targets documentation, the error argument can be used inside the `tar_target()`. If I understand correctly ` error = "continue"` would be the solution, but I am not sure that the error will propagate to all ascendants.
What are your recommendation to implementing the error handling for this scenario? Thank you.

**Comment by wlandau** [comment]
What versions of `targets` are you running? Errored targets used to return `NULL` in memory, but that was fixed in recent versions. Now, errored targets have no value at all.

> My goal with the error handling is to execute the maximum number of targets, just skip those for which an ancestor signaled an error (this workflow generates many different results, but there are common parts).

That's pretty much what `error = "continue"` does now. If a target errors, the downstream targets will try to use the most recent value from when the upstream target last ran successfully. If there is a value, that target might be skipped or it might try to run. If there is no value, the downstream target will throw an error.

**Comment by kbzsl** [comment]
Thank you for your answers.
Sorry, for ambiguous phrasing, I am not using the targets now, just I am planning to restructure an exciting application and my goal is to build the new one on top of targets. 

In the current version, in my functions the errors are handled using `try()` and a `NULL `return value in case of errors. 
I would prefer to keep the error handling inside some of the functions, but I am struggling a bit how to rewrite the error handling part in my existing code to be “compatible” with targets. Just simply replace the `return(NULL)` statement after the error catching and handling with generating a new error using the `stop()` function.

If I understood correctly what you wrote _“If a target errors, the downstream targets will try to use the most recent value from when the upstream target last ran successfully”_ –  in case of a target error, any downstream target will run, if during last run of this target produced a valid result (which can be already an invalid at the time when the error appeared).

**Reply by wlandau** [reply]
> In the current version, in my functions the errors are handled using try() and a NULL return value in case of errors.
I would prefer to keep the error handling inside some of the functions, but I am struggling a bit how to rewrite the error handling part in my existing code to be “compatible” with targets. Just simply replace the return(NULL) statement after the error catching and handling with generating a new error using the stop() function.

`targets` has its own `tryCatch()`-like error handling, and `tar_option_set(error = "continue")` allows the pipeline to continue while also still recording errors and warnings in the metadata (so you can look them up later with `tar_meta()`). I think unwrapping your existing `try()`/`tryCatch()` in your code would be ideal, but if you want to keep that, it won't prevent `targets` from working properly. You would just not be able to rely on `targets` to catch errors because you would already be doing that yourself.

> If I understood correctly what you wrote “If a target errors, the downstream targets will try to use the most recent value from when the upstream target last ran successfully” – in case of a target error, any downstream target will run, if during last run of this target produced a valid result (which can be already an invalid at the time when the error appeared).

That's right.
